src/util/walkStrategy.ts

```diff
@@ -0,0 +1,133 @@
+import { BaseTrie } from '..'
+import { FoundNodeFunction } from '../baseTrie'
+import { PrioritizedTaskExecutor } from '../prioritizedTaskExecutor'
+import { BranchNode, ExtensionNode, LeafNode, Nibbles, TrieNode } from '../trieNode'
+
+export default class WalkStrategy {
+  readonly onNode: FoundNodeFunction
+  readonly taskExecutor: PrioritizedTaskExecutor
+  readonly trie: BaseTrie
+  private resolve: Function
+
+  /**
+   * Creates a new WalkStrategy
+   * @param onNode - The `FoundNodeFunction` to call if a node is found
+   * @param trie - The `Trie` to walk on
+   */
+  private constructor(onNode: FoundNodeFunction, trie: BaseTrie, poolSize: number) {
+    this.onNode = onNode
+    this.taskExecutor = new PrioritizedTaskExecutor(poolSize)
+    this.trie = trie
+    this.resolve = () => {}
+  }
+
+  /**
+   * Async function to create and start a new walk over a trie.
+   * @param onNode - The `FoundNodeFunction to call if a node is found
+   * @param trie - The trie to walk on
+   * @param root - The root key to walk on
+   * @param poolSize - Task execution pool size to prevent OOM errors. Defaults to 500.
+   */
+  static async newWalk(
+    onNode: FoundNodeFunction,
+    trie: BaseTrie,
+    root: Buffer,
+    poolSize?: number
+  ): Promise<void> {
+    const strategy = new WalkStrategy(onNode, trie, poolSize ?? 500)
+    await strategy.startWalk(root)
+  }
+
+  private async startWalk(root: Buffer): Promise<void> {
+    return await new Promise((resolve) => {
+      this.resolve = resolve
+      this.trie
+        ._lookupNode(root)
+        .then((node) => {
+          if (!node) {
+            this.resolve()
+          } else {
+            this.processNode(root, node as TrieNode, [])
+          }
+        })
+        .catch((e) => {
+          throw e
+        })
+    })
```

Comment:
just a suggestion to simplify this with a try catch:

```suggestion
      try {
        const node = await this.trie._lookupNode(root)
        if (!node) {
          this.resolve()
        } else {
          this.processNode(root, node as TrieNode, [])
        }
      } catch (error) {
        throw error
      }
    })
```

Reply:
I did this catch thing due to the linter complaining, I think it will complain again if I try this (`no-unhandled-promises`). Will check it out.

Reply:
oh interesting, ok, not sure what part would be unhandled there, i've seen that rule a few times but usually when missing an `await`. maybe it's because it doesn't understand the `this.resolve()`?

i wonder if adding `reject` and doing `reject(e)` instead of `throw e` would help satisfy the rule in that case.

Reply:
I changed it a little bit, if you can find a better way to make the linter happy that'd be great. If I use this approach ^ then linter is not happy 😢